### PR TITLE
Send cluster specific errors from AdministratorCommissioning server

### LINKED ISSUE #11715: implement status codes in Administrator Commissioning Cluster commands
implement status codes in Administrator Commissioning Cluster commands, per Matter Spec
Screen Shot 2021-11-11 at 4 56 23 PM

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
* Fixes #11715

#### Change overview
Use cluster specific error codes on command failures in AdministratorCommissioning cluster.

#### Testing
Tested using all-cluster-app and chip-tool. The following conditions were tested.
1. Try closing commissioning window while it is not open. Expected error `EMBER_ZCL_STATUS_CODE_WINDOW_NOT_OPEN` was returned.
2. Open commissioning window. While it's open try opening commissioning window again. Expected error `EMBER_ZCL_STATUS_CODE_BUSY` was returned.
3. Try opening commissioning window with incorrect parameters. Expected error `EMBER_ZCL_STATUS_CODE_PAKE_PARAMETER_ERROR` was returned.

### CODE REVIEW

**Review by yunhanw-google — APPROVED** [review]
thanks!

**Comment by pan-apple** [comment]
Fixes #10678

File: src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp [hunk 1]
```diff
@@ -46,36 +46,41 @@ bool emberAfAdministratorCommissioningClusterOpenCommissioningWindowCallback(
     auto & salt                 = commandData.salt;
     auto & passcodeID           = commandData.passcodeID;
 
-    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    uint8_t status = 0;
```
Comment (bzbarsky-apple) [inline src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp:49]: We should use the right type here: `StatusCode`, given that we are `using namespace chip::app::Clusters::AdministratorCommissioning`.

And maybe what we should do is make this an `Optional<StatusCode>`, have the error paths do `status.Emplace(whatever)` and then down below check for `status.HasValue()`, so we are not making assumptions about 0 being a status value that can't happen....

Similar for other functions.

File: src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp [hunk 2]
```diff
@@ -46,36 +46,41 @@ bool emberAfAdministratorCommissioningClusterOpenCommissioningWindowCallback(
...
-                 status = EMBER_ZCL_STATUS_FAILURE);
-    VerifyOrExit(sizeof(verifier) == pakeVerifier.size(), status = EMBER_ZCL_STATUS_FAILURE);
-    VerifyOrExit(iterations >= kPBKDFMinimumIterations, status = EMBER_ZCL_STATUS_FAILURE);
-    VerifyOrExit(iterations <= kPBKDFMaximumIterations, status = EMBER_ZCL_STATUS_FAILURE);
-    VerifyOrExit(salt.size() >= kPBKDFMinimumSaltLen, status = EMBER_ZCL_STATUS_FAILURE);
-    VerifyOrExit(salt.size() <= kPBKDFMaximumSaltLen, status = EMBER_ZCL_STATUS_FAILURE);
-    VerifyOrExit(commissioningTimeout <= kMaxCommissionioningTimeoutSeconds, status = EMBER_ZCL_STATUS_FAILURE);
-    VerifyOrExit(discriminator <= kMaxDiscriminatorValue, status = EMBER_ZCL_STATUS_FAILURE);
+                 status = EmberAfStatusCode::EMBER_ZCL_STATUS_CODE_BUSY);
+    VerifyOrExit(sizeof(verifier) == pakeVerifier.size(), status = EmberAfStatusCode::EMBER_ZCL_STATUS_CODE_PAKE_PARAMETER_ERROR);
```
Comment (bzbarsky-apple) [inline src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp:57]:
```suggestion
    VerifyOrExit(sizeof(verifier) == pakeVerifier.size(), status = StatusCode::EMBER_ZCL_STATUS_CODE_PAKE_PARAMETER_ERROR);
```
and similar for the others.

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 3]
```diff
@@ -679,11 +679,11 @@ bool emberAfOperationalCredentialsClusterAddTrustedRootCertificateCallback(
 
     emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has added a trusted root Cert");
 
-    VerifyOrExit(gFabricBeingCommissioned.SetRootCert(RootCertificate) == CHIP_NO_ERROR, status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(gFabricBeingCommissioned.SetRootCert(RootCertificate) == CHIP_NO_ERROR, status = EMBER_ZCL_STATUS_INVALID_FIELD);
```
Comment (bzbarsky-apple) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:682]: Why this change?

Comment (pan-apple) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:682]: The spec calls for sending `INVALID_COMMAND` error and it maps to `EMBER_ZCL_STATUS_INVALID_FIELD`.

File: src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp (hunk 1 above)
Comment (pan-apple) [reply src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp:49]: Originally I tried using StatusCode as the type, but it's guarded under `#ifdef CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
`. Looks like `CHIP_USE_ENUM_CLASS_FOR_IM_ENUM` is not set. Is that not expected?

Comment (pan-apple) [reply src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp:49]:
```
#ifdef CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
// Enum for StatusCode
enum class StatusCode : uint8_t
{
    kSuccess      = 0x00,
    kBusy         = 0x01,
    kGeneralError = 0x02,
};
#else // CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
using StatusCode                           = EmberAfStatusCode;
#endif
```

Comment (bzbarsky-apple) [reply src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp:49]: So the point is, StatusCode is defined no matter what.  The question is what the values look like and whether it's an enum class or not.

If you use `StatusCode::EMBER_ZCL_STATUS_CODE_BUSY` for values and `StatusCode` to declare instances now, that reduces the amount of change we need later when we make `StatusCode` the cluster-objects enum class (to just renaming the values, not having to change the types)...

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp (hunk 3 above)
Comment (bzbarsky-apple) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:682]: Ah.  That is ... not the mapping I expected, I guess.  ;)

File: src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp [hunk 4]
```diff
@@ -46,36 +46,41 @@ bool emberAfAdministratorCommissioningClusterOpenCommissioningWindowCallback(
     auto & salt                 = commandData.salt;
     auto & passcodeID           = commandData.passcodeID;
 
-    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    Optional<StatusCode> status = Optional<StatusCode>::Missing();
```
Comment (bzbarsky-apple) [inline src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp:49]: Fwiw, Missing is the default-constructed value for Optional, so could just leave it off...

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp (hunk 4 above)
Comment (pan-apple) [reply src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp:49]: One of the example apps build was failing unless I add the Missing here. It complained about uninitialized `status` value (for example, when there is no error).

**Review by Damian-Nordic — APPROVED** [review]

File: src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply src/app/clusters/administrator-commissioning-server/administrator-commissioning-server.cpp:49]: That is ... so bizarre.  `Optional` has a default constructor!
